src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala

```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
+          val allowedFiles = Set("configuration.json") ++
+            importedEservice.descriptor.docs.map(_.path) ++
+            importedEservice.descriptor.interface.map(_.path)
+
+          val extraFiles = filePaths -- allowedFiles
+          Either.cond(
+            extraFiles.isEmpty,
+            (),
+            InvalidZipStructure(directoryName, s"Extra files found: ${extraFiles.mkString(", ")}")
+          )
+        }
+      } yield importedEservice
+    }
+
+    def deleteTempDirectory(path: Path): Unit = {
+      if (Files.isDirectory(path))
+        Files.list(path).toScala(Seq).foreach(deleteTempDirectory)
+      Files.delete(path)
+    }
+
+    def readFileAsBytes(filePath: String): Array[Byte] = {
+      Using.resource(new FileInputStream(new File(filePath))) { fileInputStream =>
+        val bytes = new Array[Byte](new File(filePath).length.toInt)
+        fileInputStream.read(bytes)
+        bytes
+      }
+    }
+
+    def verifyAndCreateImportedDoc(
+      eService: CatalogProcess.EService,
+      descriptor: CatalogProcess.EServiceDescriptor,
+      zipPath: Path,
+      file: ImportedDoc,
+      fileType: String
+    ): Future[Unit] = {
+      val filePath = zipPath.resolve(file.path).toString
+      val fileName = Paths.get(file.path).getFileName.toString
+
+      val result = for {
+        mimeType    <- Try(new Tika().detect(readFileAsBytes(filePath), fileName)).toEither.toFuture
+        contentType <- ContentType
+          .parse(mimeType)
+          .leftMap(errors => ContentTypeParsingError(mimeType, file.path, errors.map(_.toString())))
+          .toFuture
+        fileParts = (FileInfo("", fileName, contentType), new File(filePath))
+        _ <- verifyAndCreateEServiceDocument(
+          catalogProcessService,
+          fileManager,
+          eService,
+          fileParts,
+          file.prettyName,
+          fileType,
+          descriptor.id,
+          uuidSupplier
+        )
+      } yield ()
+
+      result.recoverWith { case ex: Throwable =>
+        deleteTempDirectory(zipPath)
+        for {
+          _ <- descriptor.docs.traverse { doc =>
+            catalogProcessService.deleteEServiceDocumentById(eService.id, descriptor.id, doc.id)
+          }
+          _ <- catalogProcessService.deleteEService(eService.id)
+          _ <- catalogProcessService.deleteDraft(eService.id, descriptor.id)
+        } yield throw ex
+      }
+    }
+
+    val result: Future[CreatedResource] = for {
+      tenantId         <- getOrganizationIdFuture(contexts)
+      zipFile          <- fileManager.getFile(ApplicationConfiguration.importEServiceContainer)(
+        s"${ApplicationConfiguration.importEServicePath}/$tenantId/${fileResource.filename}"
+      )
+      zipPath          <- extractZipToTempDirectory(zipFile, tenantId).toFuture
+      importedEservice <- checkZipStructure(zipPath).toFuture.recoverWith { case ex: Throwable =>
+        deleteTempDirectory(zipPath)
+        throw ex
```

Comment:
`recoverWith` expects a `Future` in return, you can use it to raise the error again.
(throwing error is avoided when using effects)
```suggestion
        Future.failed(ex)
```